asserts/ifacedecls.go

```diff
@@ -98,42 +111,94 @@ func compileMapAttrMatcher(cc compileContext, m map[string]interface{}) (attrMat
 	return matcher, nil
 }
 
-func matchEntry(apath, k string, matcher1 attrMatcher, v interface{}) error {
+func matchEntry(apath, k string, matcher1 attrMatcher, v interface{}, ctx AttrMatchContext) error {
 	apath = chain(apath, k)
-	if v == nil {
+	if _, ok := matcher1.(missingAttrMatcher); !ok && v == nil {
 		return fmt.Errorf("attribute %q has constraints but is unset", apath)
 	}
-	if err := matcher1.match(apath, v); err != nil {
+	if err := matcher1.match(apath, v, ctx); err != nil {
 		return err
 	}
 	return nil
 }
 
-func matchList(apath string, matcher attrMatcher, l []interface{}) error {
+func matchList(apath string, matcher attrMatcher, l []interface{}, ctx AttrMatchContext) error {
 	for i, elem := range l {
-		if err := matcher.match(chain(apath, strconv.Itoa(i)), elem); err != nil {
+		if err := matcher.match(chain(apath, strconv.Itoa(i)), elem, ctx); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func (matcher mapAttrMatcher) match(apath string, v interface{}) error {
+func (matcher mapAttrMatcher) match(apath string, v interface{}, ctx AttrMatchContext) error {
 	switch x := v.(type) {
 	case map[string]interface{}: // maps in attributes look like this
 		for k, matcher1 := range matcher {
-			if err := matchEntry(apath, k, matcher1, x[k]); err != nil {
+			if err := matchEntry(apath, k, matcher1, x[k], ctx); err != nil {
 				return err
 			}
 		}
 	case []interface{}:
-		return matchList(apath, matcher, x)
+		return matchList(apath, matcher, x, ctx)
 	default:
 		return fmt.Errorf("attribute %q must be a map", apath)
 	}
 	return nil
 }
 
+type missingAttrMatcher struct{}
+
+func (matcher missingAttrMatcher) match(apath string, v interface{}, ctx AttrMatchContext) error {
+	if v != nil {
+		return fmt.Errorf("attribute %q is constrained to be missing but is set", apath)
+	}
+	return nil
+}
+
+type evalAttrMatcher struct {
+	// first iteration supports just $SLOT|PLUG(argument1)
+	op  string
+	arg string
+}
+
+var (
+	validEvalAttrMatcher = regexp.MustCompile(`^\$(SLOT|PLUG)\((.+)\)$`)
```

Comment:
In the snap declarations spec, there is no reference to $SLOT and $PLUG. Are these meant to replace $same?